kvserver: add closedts side-transport consumer

andreimatei opened the pull request:
Add the consumer of closed timestamps communicated by the side transport
(i.e. the gRPC server for our new push-based streaming protocol).

This side-transport consumer accumulates closed timestamps communicated
to it by other nodes (the leaseholders of the respective ranges). Its
state is queried whenever a range needs a higher closed timestamp than
what it has locally in the Replica state, at which point the Replica's
state is lazily updated.

Release note: None
Release justification: Needed for GLOBAL tables.

andreimatei committed:
kvserver: add closedts side-transport consumer

Add the consumer of closed timestamps communicated by the side transport
(i.e. the gRPC server for our new push-based streaming protocol).

This side-transport consumer accumulates closed timestamps communicated
to it by other nodes (the leaseholders of the respective ranges). Its
state is queried whenever a range needs a higher closed timestamp than
what it has locally in the Replica state, at which point the Replica's
state is lazily updated.

Release note: None
Release justification: Needed for GLOBAL tables.

andreimatei committed:
kvserver: disable old closed timestamps mechanism

When the new one is enabled, the old one will now be disabled. No point
in running with both of them.
In order to disable the old mechanism, a new background task is added to
periodically call r.handleClosedTimestampUpdate() on all replicas with
rangefeeds - to advance the feed.

Release note: None
Release justification: Needed for GLOBAL tables.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/61305)
<!-- Reviewable:end -->

nvb reviewed (commented):
Looking good!

> Reviewed 19 of 19 files at r1, 6 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-:-MUoevUoEZoBs5j5yOX6:br2sbr6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/kv/kvserver/replica.go, line 1225 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUoYeWj0x6fHeCPH0a1:-MUoYeWj0x6fHeCPH0a2:b5q4ep8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/replica.go#L1225)):*
> ```Go
> 	}
> 
> 	var update replicaUpdate
> ```

nit: pull this above `now := `, because this is equivalent to the pattern to the `shouldExtend` stuff.
___
*[pkg/kv/kvserver/replica_follower_read.go, line 162 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUocsNT61mKRnsxob15:-MUocsNU8oZhG5N1JNT-:b-c5243w) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/replica_follower_read.go#L162)):*
> ```
> optionally takes a optional hint
> ```

Needs rewording.
___
*[pkg/kv/kvserver/replica_follower_read.go, line 171 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUodbeqElFsGbeb10fA:-MUodbeqElFsGbeb10fB:b-vnjb5x) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/replica_follower_read.go#L171)):*
> ```Go
> ) (_ hlc.Timestamp, ok bool, _ replicaUpdate) {
> 	appliedLAI := ctpb.LAI(r.mu.state.LeaseAppliedIndex)
> 	lease := *r.mu.state.Lease
> ```

nit: this is all locked, so no need to copy the lease. Let's just remove the `*`.
___
*[pkg/kv/kvserver/replica_follower_read.go, line 191 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUodqhP40G28Lf8qCt-:-MUodqhP40G28Lf8qCt0:bknpzq0) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/replica_follower_read.go#L191)):*
> ```Go
> 	}
> 
> 	if lease.Expiration != nil {
> ```

Leave a note here that we now support closed timestamps on ranges with expiration-based leases, but we'll just wait until v21.2 to avoid migration concerns.
___
*[pkg/kv/kvserver/store.go, line 668 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUoe8SPEpVwbJkQ4Nl0:-MUoe8SPEpVwbJkQ4Nl1:biqpejo) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/store.go#L668)):*
> ```Go
> 	ClosedTimestamp         *container.Container
> 	ClosedTimestampSender   *sidetransport.Sender
> 	ClosedTimestampReceiver ClosedTimestampSideTransportProvider
> ```

The interface doesn't seem useful here. Why not just `*sidetransport.Receiver`?
___
*[pkg/kv/kvserver/store.go, line 1775 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUoesZlAA-v7J68cZba:-MUoesZlAA-v7J68cZbb:boquauv) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1f1f3cc1d065cb150455ab1c2004541f7cc00f/pkg/kv/kvserver/store.go#L1775)):*
> ```Go
> }
> 
> // startClosedTimestampRangefeedSubscriber establishes a new ClosedTimestamp
> ```

Reminder that we still need to do something about this.
___
*[pkg/kv/kvserver/stores.go, line 128 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUoeIUT3lUBC8HRNcfQ:-MUoeIUT3lUBC8HRNcfR:b1bpcrg) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/stores.go#L128)):*
> ```Go
> 	ctx context.Context, rangeID roachpb.RangeID, closedTS hlc.Timestamp, lai ctpb.LAI,
> ) {
> 	if err := ls.VisitStores(func(s *Store) error {
> ```

:thumbsup: very clean.
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUoZHgp1fogJfbOU8I6:-MUoZHgp1fogJfbOU8I7:bjwfpx9) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L1)):*
> ```Go
> // Copyright 2021 The Cockroach Authors.
> ```

Rename to `receiver.go`.
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUoZNKW-6vuzJgZVb75:-MUoZNKW-6vuzJgZVb76:b-4rvmg9) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L40)):*
> ```Go
> 	mu struct {
> 		syncutil.Mutex
> 		connections map[roachpb.NodeID]*incomingStream
> ```

nit: `conns` to mirror the `Sender`.
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 63 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUoZxcR4omz7Icc-_Nt:-MUoZxcR4omz7Icc-_Nu:b-hqqrzi) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L63)):*
> ```Go
> func (s *Receiver) PushUpdates(stream ctpb.SideTransport_PushUpdatesServer) error {
> 	// Create a steam to service this connection. The stream will call back into the server
> 	// through onFirstMsg to register itself once it finds out its node id.
> ```

`s/its/the sender's`
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 64 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUoZph779GgBWdVxYpb:-MUoZph779GgBWdVxYpc:braf6cn) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L64)):*
> ```Go
> 	// Create a steam to service this connection. The stream will call back into the server
> 	// through onFirstMsg to register itself once it finds out its node id.
> 	ctx := s.AnnotateCtx(context.Background())
> ```

Doesn't `stream` have a `stream.Context()`?
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 72 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUo_7Gf5bbtMfXd9aG1:-MUo_7Gf5bbtMfXd9aG2:b-8wjjk8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L72)):*
> ```Go
> // order to use this closed timestamp.
> //
> // leaseholderNode is the last known leaseholder for the range.
> ```

Justify why we only need to look at the stream for this leaseholder.
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 104 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUo_bYFBRl63xaNoICD:-MUo_bYFBRl63xaNoICE:b-aibrxw) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L104)):*
> ```Go
> }
> 
> // onRecvErr is called when one of the inbound stream errors out. The stream is
> ```

`s/stream/streams/`
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 132 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUo_sdrC58It8OvY61G:-MUo_sdrC58It8OvY61H:b-nys58e) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L132)):*
> ```Go
> 		// lastTimestamps maintains the last (highest) closed timestamps
> 		// communicated by the client node.
> 		lastTimestamps [roachpb.MAX_CLOSED_TIMESTAMP_POLICY]hlc.Timestamp
> ```

`lastClosed` to mirror the Sender?
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 136 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUo_xOy5AYYkNs4zCUm:-MUo_xOy5AYYkNs4zCUn:b-cwxs4w) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L136)):*
> ```Go
> 		// for. These are the ranges that the closed timestamps in lastTimestamps
> 		// apply to.
> 		ranges map[roachpb.RangeID]rangeInfo
> ```

`tracked` to mirror the Sender?
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 138 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUoa2AF1GnTnh_WuexZ:-MUoa2AF1GnTnh_Wuex_:boovnxo) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L138)):*
> ```Go
> 		ranges map[roachpb.RangeID]rangeInfo
> 		// lastSeq is the sequence number of the last message received.
> 		lastSeq ctpb.SeqNum
> ```

Actually, isn't this all the same as `sender.trackedMu`? Should we pull out a shared struct. Something like `trackingState` struct that looks exactly like `sender.trackedMu` does right now.
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 142 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUo_oSJ9e92auVNLyrS:-MUo_oSJ9e92auVNLyrT:bwpmfv1) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L142)):*
> ```Go
> }
> 
> type rangeInfo struct {
> ```

Can we remove this and use `trackedRange`?
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 147 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUoaSFnET0JBsrtQa6v:-MUoaSFnET0JBsrtQa6w:b-mnos31) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L147)):*
> ```Go
> }
> 
> // StoresInterface is the Stores interface needed by the
> ```

nit: Just call this `Stores`, and leave a comment like the one on `Replica` in sender.go.
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 151 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUoaa790l1mmXNXWCiG:-MUoaa790l1mmXNXWCiH:b6t2gpz) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L151)):*
> ```Go
> type StoresInterface interface {
> 	// ForwardSideTransportClosedTimestampForRange forwards the side-transport
> 	// closed timestamp for the local replicas of the given range.
> ```

minor nit: `replica(s)` just to hint that there is usually only one replica per range.
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 164 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUoamT65OGrsPLpwdWS:-MUoamT65OGrsPLpwdWT:bfqnnjj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L164)):*
> ```Go
> }
> 
> func (r *incomingStream) String() string {
> ```

Do you still want this or was it for debugging purposes? Either way, more it below all other `incomingStream` methods - it's very distracting.
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 222 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUoayaeEs1c5dbH3ldz:-MUoayaeEs1c5dbH3le-:b2vth2g) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L222)):*
> ```Go
> 	}
> 
> 	if msg.NodeID != r.nodeID {
> ```

`else if`?
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 240 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUobdFh1N4OxLyUwdEo:-MUobdFh1N4OxLyUwdEp:b-modd2a) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L240)):*
> ```Go
> 	// Handle the removed ranges. In order to not lose closed ts info, before we
> 	// can remove a range from our tracking, we copy the info about its closed
> 	// timestamp to the local replica(s).
> ```

Add to the comment that we must do this before updating `r.mu.lastTimestamps`.
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 248 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUobQYM2V295ElpmXPn:-MUobQYM2V295ElpmXPo:b5hnpk7) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L248)):*
> ```Go
> 		r.stores.ForwardSideTransportClosedTimestampForRange(
> 			ctx, rangeID, r.mu.lastTimestamps[info.policy], info.lai)
> 		delete(r.mu.ranges, rangeID)
> ```

Pull this up before `ForwardSideTransportClosedTimestampForRange`.
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 269 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUobnW6-gM207h1F2Yw:-MUobnW6-gM207h1F2Yx:b-2k9mv0) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L269)):*
> ```Go
> 	stream ctpb.SideTransport_PushUpdatesServer,
> ) error {
> 	// We have to do the stream processing on a separate goroutine because Recv()
> ```

Is this common? I'm a little surprised we have to do this. I wonder if it's because we were ignoring the stream's context (see comment above).
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 212 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUoez9CCcqEujWj7DnC:-MUoez9CCcqEujWj7DnD:b-c0mpks) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1f1f3cc1d065cb150455ab1c2004541f7cc00f/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L212)):*
> ```Go
> // processUpdate processes one update received on the stream, updating the local
> // state.
> func (r *incomingStream) processUpdate(ctx context.Context, msg *ctpb.Update) {
> ```

This seems straightforward to unit test. We'll just need to mock out the `Stores` interface.
___
*[pkg/kv/kvserver/closedts/sidetransport/sender.go, line 190 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUocUt04E3IsumbzEPg:-MUocUt04E3IsumbzEPh:b-3q5wga) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/sender.go#L190)):*
> ```Go
> 				} else {
> 					// Disable the side-transport.
> 					timer.Stop()
> ```

Reminder: I think you're fixing something here right now.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61305)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @nvanbenschoten)

___
*[pkg/kv/kvserver/replica.go, line 1225 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUoYeWj0x6fHeCPH0a1:-MUpLZFCCuWwYA8SEnJy:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/replica.go#L1225)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: pull this above `now := `, because this is equivalent to the pattern to the `shouldExtend` stuff.
</blockquote></details>

done
___
*[pkg/kv/kvserver/replica_follower_read.go, line 162 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUocsNT61mKRnsxob15:-MUpLdEsDFyuQT_lU0SB:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/replica_follower_read.go#L162)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> ```
> optionally takes a optional hint
> ```

Needs rewording.
</blockquote></details>

done
___
*[pkg/kv/kvserver/replica_follower_read.go, line 171 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUodbeqElFsGbeb10fA:-MUpLjmkDIUKNLygqtqd:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/replica_follower_read.go#L171)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: this is all locked, so no need to copy the lease. Let's just remove the `*`.
</blockquote></details>

done
___
*[pkg/kv/kvserver/replica_follower_read.go, line 191 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUodqhP40G28Lf8qCt-:-MUpMBTw49PDf8IfXUls:brtvhz0) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/replica_follower_read.go#L191)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Leave a note here that we now support closed timestamps on ranges with expiration-based leases, but we'll just wait until v21.2 to avoid migration concerns.
</blockquote></details>

done. I had hesitated before because of concerns that even longer term these expiration-based ranges don't necessarily have a lease - but then again neither do the regular ranges necessarily.
___
*[pkg/kv/kvserver/store.go, line 668 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUoe8SPEpVwbJkQ4Nl0:-MUpMgkNEFag2P0gQdqa:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/store.go#L668)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

The interface doesn't seem useful here. Why not just `*sidetransport.Receiver`?
</blockquote></details>

done
___
*[pkg/kv/kvserver/store.go, line 1775 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUoesZlAA-v7J68cZba:-MUpDpcD9u_485D9Fzyd:b-iw91j3) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1f1f3cc1d065cb150455ab1c2004541f7cc00f/pkg/kv/kvserver/store.go#L1775)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Reminder that we still need to do something about this.
</blockquote></details>

see now, in the 2nd commit
___
*[pkg/kv/kvserver/closedts/sidetransport/sender.go, line 190 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUocUt04E3IsumbzEPg:-MUpDb7F734QMEja5D_W:b-zg06z) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/sender.go#L190)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Reminder: I think you're fixing something here right now.
</blockquote></details>

https://github.com/cockroachdb/cockroach/pull/61373
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUoZHgp1fogJfbOU8I6:-MUpFk1F35WYNPAv51yv:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L1)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Rename to `receiver.go`.
</blockquote></details>

done
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUoZNKW-6vuzJgZVb75:-MUpFu5j54yttu8iuWZE:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L40)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: `conns` to mirror the `Sender`.
</blockquote></details>

done
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 63 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUoZxcR4omz7Icc-_Nt:-MUpG5Q11bA9-RY2d1Qc:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L63)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`s/its/the sender's`
</blockquote></details>

done
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 64 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUoZph779GgBWdVxYpb:-MUpG8T_CbyR_UEDiyZO:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L64)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Doesn't `stream` have a `stream.Context()`?
</blockquote></details>

done
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 72 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUo_7Gf5bbtMfXd9aG1:-MUpGQiS7DRSFCQW_3Lg:b-2x4j6k) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L72)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Justify why we only need to look at the stream for this leaseholder.
</blockquote></details>

see now
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 104 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUo_bYFBRl63xaNoICD:-MUpGVXAAWpuv6lTnnnv:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L104)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`s/stream/streams/`
</blockquote></details>

done
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 132 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUo_sdrC58It8OvY61G:-MUpGazB2jOEWpvPOER6:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L132)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`lastClosed` to mirror the Sender?
</blockquote></details>

done
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 136 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUo_xOy5AYYkNs4zCUm:-MUpJgmR5q_3_G77W1Dl:b2ni76j) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L136)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`tracked` to mirror the Sender?
</blockquote></details>

n/a
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 138 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUoa2AF1GnTnh_WuexZ:-MUpJhbs1tDbp6duaZY-:bpyqtb6) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L138)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Actually, isn't this all the same as `sender.trackedMu`? Should we pull out a shared struct. Something like `trackingState` struct that looks exactly like `sender.trackedMu` does right now.
</blockquote></details>

Done. I think I like.
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 142 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUo_oSJ9e92auVNLyrS:-MUpJoXoAGJM4A-275s9:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L142)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Can we remove this and use `trackedRange`?
</blockquote></details>

done
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 147 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUoaSFnET0JBsrtQa6v:-MUpJylxDtddaXCK-kxK:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L147)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: Just call this `Stores`, and leave a comment like the one on `Replica` in sender.go.
</blockquote></details>

done
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 151 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUoaa790l1mmXNXWCiG:-MUpK4QxBcFztavEQnbb:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L151)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

minor nit: `replica(s)` just to hint that there is usually only one replica per range.
</blockquote></details>

done
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 164 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUoamT65OGrsPLpwdWS:-MUpDwn85S5c2vGwUBE2:b-z9dcp2) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L164)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do you still want this or was it for debugging purposes? Either way, more it below all other `incomingStream` methods - it's very distracting.
</blockquote></details>

still want it, although it's not yet used.
Moved.
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 222 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUoayaeEs1c5dbH3ldz:-MUpE89J5zOy6lfma_GG:b-qtwgl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L222)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`else if`?
</blockquote></details>

mmm I don't know. If it was a panic instread of `log.Fatal()` above, the linter wouldn't even let me do `else`.
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 240 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUobdFh1N4OxLyUwdEo:-MUpKSe67PNYPlgRxCPC:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L240)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Add to the comment that we must do this before updating `r.mu.lastTimestamps`.
</blockquote></details>

done
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 248 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUobQYM2V295ElpmXPn:-MUpKYNHBOFY-ztKhaBk:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L248)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Pull this up before `ForwardSideTransportClosedTimestampForRange`.
</blockquote></details>

done
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 269 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUobnW6-gM207h1F2Yw:-MUpKbMDE7Aj4Unpzp4W:b-hi5ebn) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7edbe4451d3feee8579bc4efbdc60ebdfd84e18/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L269)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is this common? I'm a little surprised we have to do this. I wonder if it's because we were ignoring the stream's context (see comment above).
</blockquote></details>

I was surprised too and spent an hour in old, frustrating gRPC threads. The only way to interrupt a Recv() is to return from the RPC handler. I'm not sure under what circumstances the stream's context gets canceled, but it certainly does not happen in direct relation with our stopper quiescence. It's possible that it happens in connection to calling some type of drain/stop on the gRPC server, but I dunno.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
Looking good. I think we'll still want a few basic unit tests right around `incomingStream.processUpdate`, but after that and once CI is happy, I think we can merge.

> Reviewed 13 of 16 files at r3, 11 of 11 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-:-MUq-XQ02Pl-g3xKlsGN:blh2j53)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/kv/kvserver/closedts/sidetransport/sender.go, line 37 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUq-DMc2Yegw0xisJOO:-MUq-DMc2Yegw0xisJOP:b-vsfjc9) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e5a2f235bf427c8c8ed7c7827704dc2e921ec49/pkg/kv/kvserver/closedts/sidetransport/sender.go#L37)):*
> ```Go
> 
> // Sender represents the sending-side of the closed timestamps "side-transport".
> // Its role is to periodically advance the closed timestamps of all the tracked
> ```

Looks like a rename from `tracked` to `ranges` went bad all over the file.
___
*[pkg/kv/kvserver/closedts/sidetransport/sender_test.go, line 194 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUq-TMh6irKQCvyhv-e:-MUq-TMh6irKQCvyhv-f:brc82bk) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e5a2f235bf427c8c8ed7c7827704dc2e921ec49/pkg/kv/kvserver/closedts/sidetransport/sender_test.go#L194)):*
> ```Go
> 
> func TestSenderConnectionChanges(t *testing.T) {
> 	// TODO: Two tracked.
> ```

Rename went bad here too.


<!-- Sent from Reviewable.io -->

andreimatei committed:
closedts/sidetransport: speed up a test

Make a test send larger proto messages, to fill network buffers fasters.
Makes a difference under race.

Release note: None
Release justification: Tests only

andreimatei reviewed (commented):
Fixed the deadlock between removals and replicas querying the closed timestamp.
Added some tests.
Moved the Receiver to an RWMutex.
Appended a commit mucking with a test.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-:-MUzXwie5hpUTHjBrakg:b-xljgu9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @nvanbenschoten)

___
*[pkg/kv/kvserver/closedts/sidetransport/sender.go, line 37 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUq-DMc2Yegw0xisJOO:-MUupRv259npwHIqPzrQ:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e5a2f235bf427c8c8ed7c7827704dc2e921ec49/pkg/kv/kvserver/closedts/sidetransport/sender.go#L37)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Looks like a rename from `tracked` to `ranges` went bad all over the file.
</blockquote></details>

fixed
___
*[pkg/kv/kvserver/closedts/sidetransport/sender_test.go, line 194 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUq-TMh6irKQCvyhv-e:-MUuqipJCD-0ZjHLwdFP:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e5a2f235bf427c8c8ed7c7827704dc2e921ec49/pkg/kv/kvserver/closedts/sidetransport/sender_test.go#L194)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Rename went bad here too.
</blockquote></details>

fixed


<!-- Sent from Reviewable.io -->

nvb reviewed (changes requested):
> Reviewed 8 of 19 files at r5, 11 of 11 files at r6, 2 of 2 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61305)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/kv/kvserver/closedts/sidetransport/sender_test.go, line 52 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUznCWq74qOMKPx-OGu:-MUznCWq74qOMKPx-OGv:b-6ed94o) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dda86376f10e6cf23e4b44b5a395d52596db5ab/pkg/kv/kvserver/closedts/sidetransport/sender_test.go#L52)):*
> ```Go
> func (m *mockReplica) StoreID() roachpb.StoreID    { return m.storeID }
> func (m *mockReplica) GetRangeID() roachpb.RangeID { return m.rangeID }
> func (m *mockReplica) Desc() *roachpb.RangeDescriptor {
> ```

This has me wondering whether we should return the descriptor from `BumpSideTransportClosed` to avoid locking each replica twice. What do you think? Doesn't need to be in this change.
___
*[pkg/kv/kvserver/closedts/sidetransport/sender_test.go, line 505 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUzmzxeDUd67bcYJvxy:-MUzmzxeDUd67bcYJvxz:b-niff45) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dda86376f10e6cf23e4b44b5a395d52596db5ab/pkg/kv/kvserver/closedts/sidetransport/sender_test.go#L505)):*
> ```Go
> 	// 1) and yet that Receiver doesn't leak goroutine, which proves that nobody
> 	// connected to it (a node doesn't connect to itself).
> 	go func() {
> ```

Is this a goroutine because we don't expect the receive to complete until after the stopper has been triggered?
___
*[pkg/kv/kvserver/closedts/sidetransport/receiver.go, line 148 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUzlg-f4u8s_rykRw7U:-MUzlg-g896ga3JkRMtd:b-kongwq) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dda86376f10e6cf23e4b44b5a395d52596db5ab/pkg/kv/kvserver/closedts/sidetransport/receiver.go#L148)):*
> ```Go
> 	}
> }
> type incomingStreamTestingKnobs struct {
> ```

nit: spacing above
___
*[pkg/kv/kvserver/closedts/sidetransport/receiver.go, line 262 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUzlvlUElzjdcCe72YH:-MUzlvlUElzjdcCe72YI:b-qxgin4) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dda86376f10e6cf23e4b44b5a395d52596db5ab/pkg/kv/kvserver/closedts/sidetransport/receiver.go#L262)):*
> ```Go
> 	for i, rangeID := range removed {
> 		info := infos[i]
> 		r.stores.ForwardSideTransportClosedTimestampForRange(
> ```

Why can't we do this under the read lock? I thought that was part of the point? It would be a shame if we needed to allocate the `infos` slice for ranges that aren't even local.

Maybe this structure was built before the RWMutex split.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61305)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @nvanbenschoten)

___
*[pkg/kv/kvserver/closedts/sidetransport/sender_test.go, line 52 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUznCWq74qOMKPx-OGu:-MV2C-mj-jh6imvnjHmp:b27f4gg) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dda86376f10e6cf23e4b44b5a395d52596db5ab/pkg/kv/kvserver/closedts/sidetransport/sender_test.go#L52)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This has me wondering whether we should return the descriptor from `BumpSideTransportClosed` to avoid locking each replica twice. What do you think? Doesn't need to be in this change.
</blockquote></details>

Good idea. Will do separately.
___
*[pkg/kv/kvserver/closedts/sidetransport/sender_test.go, line 505 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUzmzxeDUd67bcYJvxy:-MV2JVQh5LD7DQkYS98G:bry1hq2) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dda86376f10e6cf23e4b44b5a395d52596db5ab/pkg/kv/kvserver/closedts/sidetransport/sender_test.go#L505)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is this a goroutine because we don't expect the receive to complete until after the stopper has been triggered?
</blockquote></details>

yes. But this was pretty ugly. See now.
___
*[pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go, line 212 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUoez9CCcqEujWj7DnC:-MV299tK7G5JZ0kg2Dz_:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1f1f3cc1d065cb150455ab1c2004541f7cc00f/pkg/kv/kvserver/closedts/sidetransport/closed_timestamp_receiver.go#L212)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This seems straightforward to unit test. We'll just need to mock out the `Stores` interface.
</blockquote></details>

done
___
*[pkg/kv/kvserver/closedts/sidetransport/receiver.go, line 148 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUzlg-f4u8s_rykRw7U:-MV298Xc89QqOVgs5ndi:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dda86376f10e6cf23e4b44b5a395d52596db5ab/pkg/kv/kvserver/closedts/sidetransport/receiver.go#L148)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: spacing above
</blockquote></details>

done
___
*[pkg/kv/kvserver/closedts/sidetransport/receiver.go, line 262 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUzlvlUElzjdcCe72YH:-MV2AInx08QA4eENk0Y6:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dda86376f10e6cf23e4b44b5a395d52596db5ab/pkg/kv/kvserver/closedts/sidetransport/receiver.go#L262)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why can't we do this under the read lock? I thought that was part of the point? It would be a shame if we needed to allocate the `infos` slice for ranges that aren't even local.

Maybe this structure was built before the RWMutex split.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 15 files at r8, 11 of 11 files at r9, 2 of 2 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-:-MV2P5Zq2GtMrfEAaGRQ:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/kv/kvserver/closedts/sidetransport/sender_test.go, line 52 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUznCWq74qOMKPx-OGu:-MV2P-8Z1faGcyfggg0p:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dda86376f10e6cf23e4b44b5a395d52596db5ab/pkg/kv/kvserver/closedts/sidetransport/sender_test.go#L52)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Good idea. Will do separately.
</blockquote></details>

:thumbsup: 
___
*[pkg/kv/kvserver/closedts/sidetransport/receiver.go, line 262 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUzlvlUElzjdcCe72YH:-MV2OR2b6AG2nODmV9tB:b-oprp5z) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dda86376f10e6cf23e4b44b5a395d52596db5ab/pkg/kv/kvserver/closedts/sidetransport/receiver.go#L262)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

done
</blockquote></details>

I think we can still do a bit better. Now that this is two phases, let's merge the `delete(r.mu.tracked, rangeID)` critical section with the main critical section in `processUpdate`. So then it's one RLock and one Lock. Up to you whether you want to still keep this method for the RLock part (but either way, the `if len(removed) > 0 {` check is :thumbsup: )


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61305)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei and @nvanbenschoten)

___
*[pkg/kv/kvserver/closedts/sidetransport/receiver.go, line 262 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUzlvlUElzjdcCe72YH:-MV2TUKk100Waem62PQM:b-2x4j6k) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dda86376f10e6cf23e4b44b5a395d52596db5ab/pkg/kv/kvserver/closedts/sidetransport/receiver.go#L262)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I think we can still do a bit better. Now that this is two phases, let's merge the `delete(r.mu.tracked, rangeID)` critical section with the main critical section in `processUpdate`. So then it's one RLock and one Lock. Up to you whether you want to still keep this method for the RLock part (but either way, the `if len(removed) > 0 {` check is :thumbsup: )
</blockquote></details>

see now


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 16 files at r11, 11 of 11 files at r12, 2 of 2 files at r13.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-:-MV2V4cd-2V2wpXCsyOP:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/kv/kvserver/closedts/sidetransport/receiver.go, line 262 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/61305#-MUzlvlUElzjdcCe72YH:-MV2V28UCj3tZfihNh3W:bdeavht) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dda86376f10e6cf23e4b44b5a395d52596db5ab/pkg/kv/kvserver/closedts/sidetransport/receiver.go#L262)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

see now
</blockquote></details>

:fire: 


<!-- Sent from Reviewable.io -->

andreimatei commented:
bors r+

awoods187 commented:
Congrats on the merge!
